File: src/main/scala/kim/dencka/java/easy/collection/BinarySearch.java
package kim.dencka.java.easy.collection;

class BinarySearch {
    public static void main(String[] args) {
        int[] input = {-1, 0, 3, 5, 9, 12};
        System.out.println(search(input, 9));
        System.out.println(search(input, 2));
    }

    public static int search(int[] nums, int target) {
        int low = 0;
        int high = nums.length - 1;
        int mid;
        while (low <= high) {
            mid = low + (high - low) / 2;
            if (nums[mid] > target) high = mid - 1;
            else if (nums[mid] < target) low = mid + 1;
            else return mid;
        }
        return -1;
    }
}

File: src/main/scala/kim/dencka/scala/easy/BinarySearch.scala
package kim.dencka.scala.easy

object BinarySearch extends App {

  def search(nums: Array[Int], target: Int): Int = {
    @annotation.tailrec
    def findNumber(low: Int, high: Int): Int = {
      val middle = low + (high - low) / 2

      if (low > high) -1
      else if (nums(middle) == target) middle
      else if (nums(middle) > target) findNumber(low, middle - 1)
      else findNumber(middle + 1, high)
    }

    findNumber(0, nums.length - 1)
  }

  println(search(Array(-1, 0, 3, 5, 9, 12), 2))

}
